### PR TITLE
Implement `fn` for `Real`

### PR DESCRIPTION
**Pull request by samestep**
This PR generalizes the functionality from `packages/core/src/translate.test.ts` into initial implementations of `fn` and `interp` functions which can be used to construct Rose functions by using a normal JavaScript function to construct a computation DAG, convert that to Rose IR, and call the interpreter as before.

I created a new `struct Body` in `rose-web` to hold the list of instructions while it is being constructed, because I realized that all we're doing is pushing instructions onto the end, so it's easiest to just define a method for each kind of instruction and call those in the implementation of `fn`, rather than using the ts-rs `Instr` type followed by Serde.

Incidental change: I enabled [`noFallthroughCasesInSwitch`](https://www.typescriptlang.org/tsconfig#noFallthroughCasesInSwitch) in `packages/core/tsconfig.json` because `strict` doesn't automatically enable it, and it's useful for catching bugs in the kind of `switch` statements we're using in `packages/core/src/fn.ts`.

### CODE REVIEW

File: packages/core/src/interp.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import * as ffi from "./ffi.js";
+import { Fn } from "./fn.js";
+import { Real } from "./real.js";
+
+type Resolve<T extends readonly Real[]> = {
+  [K in keyof T]: number;
+};
+
```
Comment (ravenrothkopf) [inline packages/core/src/interp.ts:8]: `interp` needs some comments

File: packages/core/src/fn.ts [hunk 2]
```diff
@@ -1,22 +1,157 @@
...
+    case "Int":
+    case "Real": {
+      return t.tag;
+    }
+    case "Vec": {
+      throw Error("TODO");
+    }
+  }
+};
+
```
Comment (ravenrothkopf) [inline packages/core/src/fn.ts:165]: def put some comments on `fn`!

Comment (ravenrothkopf) [reply packages/core/src/fn.ts:165]: like how we're constructing `g` and stuff

File: packages/core/src/ffi.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import * as wasm from "@rose-lang/wasm";
+import { Type } from "@rose-lang/wasm/core/Type";
+import { Val } from "@rose-lang/wasm/interp/Val";
+
```
Comment (ravenrothkopf) [inline packages/core/src/ffi.ts:4]: I think it would be good to comment a bit about garbage collection here

File: packages/core/src/fn.ts [hunk 4]
```diff
@@ -1,22 +1,157 @@
...
+  }
+
+  local(x: unknown): number {
+    const id = this.locals.size;
+    this.locals.set(x, id);
+    this.instrs.set(id);
+    this.instrs.get(id);
+    return id;
+  }
+
```
Comment (ravenrothkopf) [inline packages/core/src/fn.ts:81]: kind confused about what's going on here with `val<T>`, is this for branching cases?

**Review by ravenrothkopf — APPROVED** [review]
The code looks good to me, although it would be nice to walk through it together if we have time! I think just add more comments like everywhere lol, especially some higher-level comments explaining the relationship/interactions between `wasm` and `js`

Overall super great progress tho!!

File: packages/core/src/fn.ts (hunk 4 above)
Comment (samestep) [reply packages/core/src/fn.ts:81]: Yeah, `Val` is for kinds of nodes that can appear regardless of type, such as `symbol`s representing function parameters, or `Call`s to other functions, or `Index`ing into a vector, or `If`/else conditionals.

**Review by ravenrothkopf — APPROVED** [review]
LGTM

**Pull request merged**